File: src/main/java/com/github/katorly/starlinutils/StarlinUtils.java
package com.github.katorly.starlinutils;

import java.text.ParseException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import com.github.katorly.starlinutils.backup.ConfigReader;
import com.github.katorly.starlinutils.commands.*;
import com.github.katorly.starlinutils.utils.PlayTime;

import org.bukkit.Bukkit;
import org.bukkit.NamespacedKey;
import org.bukkit.entity.Player;
import org.bukkit.event.HandlerList;
import org.bukkit.plugin.java.JavaPlugin;

public class StarlinUtils extends JavaPlugin {
    public static StarlinUtils INSTANCE;
    public StarlinUtils() {
        INSTANCE = this;
    }

    public static ConfigReader config;
    public static ConfigReader timedata;
    public static ConfigReader monthly;
    public Map<UUID, Long> StartTime = new HashMap<>();
    public static boolean serverClosing = false;
    public final List<NamespacedKey> recipeKeys = new ArrayList<>();
    // public static ProtocolManager cab = ProtocolLibrary.getProtocolManager(); //清明节

    @Override
    public void onEnable() {
        getServer().getPluginManager().registerEvents(new EventListener(), this);
        // ClearAndBright.grayChatting(); //清明节
        // getServer().getPluginManager().registerEvents(new AprilFools(),this); //愚人节
        config = new ConfigReader(this,"","config.yml");
	    config.saveDefaultConfig();
        timedata = new ConfigReader(this, "", "timedata.yml");
        timedata.saveDefaultConfig();
        monthly = new ConfigReader(this, "", "monthly.yml");
        monthly.saveDefaultConfig();
        Bukkit.getPluginCommand("su").setExecutor(new su());
        Bukkit.getPluginCommand("su").setTabCompleter(new su());
        Bukkit.getPluginCommand("givefly").setExecutor(new givefly());
        Bukkit.getPluginCommand("givefly").setTabCompleter(new givefly());
        Bukkit.getPluginCommand("listfly").setExecutor(new listfly());
        Bukkit.getPluginCommand("delfly").setExecutor(new delfly());
        Bukkit.getPluginCommand("delfly").setTabCompleter(new delfly());
        Recipe.registerConcreteRecipe(); //Add colored concrete recipe
        Bukkit.getLogger().info("[StarlinUtils] Repo: https://github.com/katorlys/StarlinUtils");
        Bukkit.getLogger().info("[StarlinUtils] StarlinUtils enabled! Made for StarlinWorld server only.");
        for (Player p : Bukkit.getOnlinePlayers()) {
            try {
                PlayTime.initialize(p);
            } catch (ParseException e) {
                Bukkit.getLogger().severe("[StarlinUtils] Error counting player's monthly play time.");
                e.printStackTrace();
            }
        }
    }

    @Override
    public void onDisable() {
        for (Player p : Bukkit.getOnlinePlayers()) {
            try {
                PlayTime.settle(p);
            } catch (ParseException e) {
                Bukkit.getLogger().severe("[StarlinUtils] Error counting player's monthly play time.");
                e.printStackTrace();
            }
        }
        recipeKeys.forEach(Bukkit::removeRecipe);
        recipeKeys.clear();
        HandlerList.unregisterAll(this);
        config.reloadConfig();
        timedata.reloadConfig();
        monthly.reloadConfig();
        Bukkit.getLogger().info("[StarlinUtils] StarlinUtils disabled!");
    }
}

File: src/main/kotlin/com/github/katorly/starlinutils/StarlinUtils.kt
/*
 * Copyright (c) 2020-2023 Katorly Lab (https://github.com/katorlys)
 *
 * This Source Code Form is licensed under CC BY-NC-ND 4.0
 * (Attribution-NonCommercial-NoDerivatives 4.0
 * International). To view a copy of this license, visit
 * http://creativecommons.org/licenses/by-nc-nd/4.0/.
 */

package com.github.katorly.starlinutils

import com.github.katorly.starlinutils.ConfigHandler.conf
import com.github.katorly.starlinutils.ConfigHandler.reloadConfig
import com.github.katorly.starlinutils.tools.PlayTime.initPlayTime
import com.github.katorly.starlinutils.tools.PlayTime.settlePlayTime
import org.bukkit.Bukkit
import org.bukkit.Bukkit.getScheduler
import org.bukkit.NamespacedKey
import org.bukkit.event.HandlerList.unregisterAll
import taboolib.common.platform.Plugin
import taboolib.common.platform.function.getDataFolder
import taboolib.common.platform.function.info
import taboolib.common.platform.function.submitAsync
import taboolib.expansion.setupPlayerDatabase
import taboolib.platform.BukkitPlugin
import java.io.File


object StarlinUtils : Plugin() {
    val plugin by lazy { BukkitPlugin.getInstance() }
    var serverClosing = false
    val recipeKeys: MutableList<NamespacedKey> = ArrayList()

    override fun onEnable() {
        reloadConfig()
        if (conf.getBoolean("playtime.enable-mysql")) {
            setupPlayerDatabase(conf.getConfigurationSection("playtime.mysql")!!)
        } else {
            setupPlayerDatabase(File(getDataFolder(), "data.db"))
        }
        RecipeHandler.registerConcreteRecipe()
        if (plugin.server.pluginManager.getPlugin("LuckPerms") != null) info("[StarlinUtils] 检测到 LuckPerms, 现在可以使用 /setupfly 指令来初始化飞行权限组 (如果还没初始化的话).")
        info("[StarlinUtils] 仓库: https://github.com/katorlys/StarlinUtils")
        info("[StarlinUtils] StarlinUtils 已加载! 为 星林宇宙 而做.")

        // PlayTime 相关功能
        info("[StarlinUtils] 初始化在线玩家游玩时间中......")
        submitAsync(true) {
            if (Bukkit.getOnlinePlayers().isNotEmpty()) {
                Bukkit.getOnlinePlayers().forEach {
                    initPlayTime(it)
                }
            }
        }
        info("[StarlinUtils] 初始化在线玩家游玩时间完成!")
    }

    override fun onDisable() {
        // PlayTime 相关功能
        info("[StarlinUtils] 保存玩家游玩时间中......")
        if (Bukkit.getOnlinePlayers().isNotEmpty()) {
            Bukkit.getOnlinePlayers().forEach {
                settlePlayTime(it)
            }
        }
        info("[StarlinUtils] 保存玩家游玩时间完成!")

        reloadConfig()
        unregisterAll()
        getScheduler().cancelTasks(plugin)
        recipeKeys.forEach(Bukkit::removeRecipe)
        recipeKeys.clear()
        info("[StarlinUtils] StarlinUtils 已卸载!")
    }
}